### DISCUSSION TITLE
Tagged String - "Ambiguous use of "init(_:)'"

### DISCUSSION THREAD

**Discussion by philipbel** [pointfreeco/swift-tagged#76] [General]
Consider the following:
```Swift
enum _StringTag { }
typealias MyString = Tagged<_StringTag, String>

let aString = "bar"
let myString = MyString(aString)
```

Compiler complains that the `MyString.init()` call is ambiguous. I would expect this to just work, since I am passing a `String` where `RawValue == String`.
```
[...]:18:16: error: ambiguous use of 'init(_:)'
let myString = MyString(aString)
               ^
[...]/Tagged.swift:9:10: note: found this candidate
  public init(_ rawValue: RawValue) {
         ^
[...]/Tagged.swift:199:10: note: found this candidate
  public init?(_ description: String) {
```

And indeed, from `Tagged.swift`:
```Swift
[...]
  public init(_ rawValue: RawValue) {
    self.rawValue = rawValue
  }
[...]
extension Tagged: LosslessStringConvertible where RawValue: LosslessStringConvertible {
  public init?(_ description: String) {
    guard let rawValue = RawValue(description) else { return nil }
    self.init(rawValue: rawValue)
  }
}
```

**Comment by mbrandonw** [comment]
Hi @philipbel, this is just an unfortunate consequence of trying to support a succinct initialization syntax (e.g. `Tagged(value)`), and some conveniences, like `LosslessStringConvertible`.

But at the end of the day you can always disambiguate by just using the more verbose `Tagged.init(rawValue:)`:

https://github.com/pointfreeco/swift-tagged/blob/8ce7c8c800372cc1f7e153e55b79127132ed82e4/Sources/Tagged/Tagged.swift#L5-L7

There is also chance using `@_disfavoredOverload` on the `LosslessStringConvertible` initializer would have also helped. If you want to PR that we will consider adding it to the library.

But since this isn't really an issue with the library I am going to convert it to a discussion. please feel free to continue the conversation over there.
